ml4gw/dataloading/chunked_dataset.py

```diff
@@ -0,0 +1,267 @@
+from typing import List
+
+import h5py
+import numpy as np
+import torch
+
+
+class ChunkLoader(torch.utils.data.IterableDataset):
+    def __init__(
+        self,
+        fnames: List[str],
+        channels: List[str],
+        chunk_size: int,
+        reads_per_chunk: int,
+        chunks_per_epoch: int,
+        coincident: bool = True,
+    ) -> None:
+        self.fnames = fnames
+        self.channels = channels
+        self.chunk_size = chunk_size
+        self.reads_per_chunk = reads_per_chunk
+        self.chunks_per_epoch = chunks_per_epoch
+        self.coincident = coincident
+
+        sizes = []
+        for f in self.fnames:
+            with h5py.File(f, "r") as f:
+                size = len(f[self.channels[0]])
+                sizes.append(size)
+        total = sum(sizes)
+        self.probs = np.array([i / total for i in sizes])
+
+    def sample_fnames(self):
+        return np.random.choice(
+            self.fnames,
+            p=self.probs,
+            size=(self.reads_per_chunk,),
+            replace=True,
+        )
+
+    def load_coincident(self):
+        fnames = self.sample_fnames()
+        chunks = []
+        for fname in fnames:
+            with h5py.File(fname, "r") as f:
+                chunk, idx = [], None
+                for channel in self.channels:
+                    if idx is None:
+                        end = len(f[channel]) - self.chunk_size
+                        idx = np.random.randint(0, end)
```

Comment:
Could be cleaner to have a single load function and do something like
```
for channel in self.channels:
    end = len(f[channel]) - self.chunk_size
    if idx is None or not coincident: 
        idx = np.random.randint(0, end)
    x = f[channel][idx : idx + self.chunk_size]

```

Reply:
Yeah been thinking of how to combine these but the issue is that for non-coincident sampling, you kind of flip the order: instead of sampling a bunch of files up front and then reading each channel from them, you loop through the _channels_ then sample a different set of files for each one.

Reply:
Ah, yeah, I see, I hadn't noticed the change in order between the two functions. What about something like this?

```
def load(self):
        chunks = []
        for channel in self.channels:
            if channel == self.channels[0] or not coincident:
                 fnames = self.sample_fnames()
            chunk, idx = [], None
            for fname in fnames:
                with h5py.File(fname, "r") as f:
                    end = len(f[channel]) - self.chunk_size
                    if idx is None or not coincident:
                        idx = np.random.randint(0, end)
                    x = f[channel][idx : idx + self.chunk_size]
                    chunk.append(x)
            chunks.append(np.stack(chunk))
        return np.stack(chunks, axis=1)
```

Reply:
Ok yeah I like this. The one issue is linters will complain that `fnames` is only defined under an `if` statement, and they have no way of knowing that the `if` will _always_ be triggered and so will warn of potential `NameError`s, but that's an easy enough fix (that's why I do the dummy `idx = None`). I'll get this implemented ASAP

Reply:
FWIW, tried to implement this but realized there's an issue with the `idx`: you need to keep the same indices for each channel when you sample coincidently, but this implementation samples new indices for each channel within a given file. Played around with making this work via `dict`s, but ultimately became more onerous and less readable than just implementing as two separate functions, but happy to take a look if you have a clean implementation.

Reply:
How about defining `self.sizes = sizes` up in `__init__`? And then here you could do

```
chunks, fnames = [], None
for channel in self.channels:
    if fnames is None or not self.coincident:
        fnames = self.sample_fnames()
        sizes = self.sizes[np.argwhere(self.fnames == fnames)]
        ends = [size - self.chunk_size for size in sizes]
        idxs = [np.random.randint(0, end) for end in ends]
    chunk = []
    for idx, fname in zip(idxs, fnames):
        with h5py.File(fname, "r") as f:
            x = f[channel][idx : idx + self.chunk_size]
            chunk.append(x)
        chunks.append(np.stack(chunk))
return np.stack(chunks, axis=1)
```

Reply:
Did you get a chance to take a look at this?

Reply:
For the record: Will and I chatted over Slack about the downside of this approach being that you're doing unnecessary extra file openings inside the loop for the `coincident=True` case, which can introduce non-trivial latency, especially for shorter chunks. Probably easier to keep these separate for now, but also raises the question of whether we should introduce a parameter for the user to control whether non-coincident sampling reads from independent files or from independently sampled times from the _same_ file to avoid this latency as well. Will file an issue for this with the relevant latency data for reference.